### DISCUSSION TITLE
Правила именования в БЭМ

### DISCUSSION THREAD

**Discussion by furtivite** [doka-guide/platform#278] [Q&A] [answered]
В вёрстке мы пришли к решению использовать БЭМ как правила в инфраструктуре и оформлении кода. 

Необходимо прийти к решению о стилях оформления кода:

- [Вариант принятый в Яндексе](https://ru.bem.info/methodology/naming-convention/)
- [Альтернативный вариант Two Dashes](https://ru.bem.info/methodology/naming-convention/#стиль-two-dashes)

и другие. 

Присоединитесь, пожалуйста, в обсуждение, чтобы мы пришли к единому мнению. Спасибо.

**Comment by lenaryan** [comment]
Для меня удобнее вариант с двумя дефисами - он визуально более понятный. Всё-таки дефис от нижнего подчёркивания отличить проще, чем два нижних подчёркивания от одного.

**Reply by pepelsbey** [reply]
Я тоже за международный вариант с двумя дефисами — он больше распространён в статьях, в документации, в коде проектов. Ну и лучше читается. Но я бы тут дождался @monochromer с примерами ключ-значение, чтобы понять, не будет ли он нам жать.

**Reply by monochromer** [reply]
Вариант `theme_color_light` для меня понятнее, чем `theme--color--light`

**Reply by furtivite** [reply]
От `theme--color--light` появляется ощущение элемента элемента, который БЭМом не поддерживается.

**Reply by pepelsbey** [reply]
Ну неет, `theme--color--light` это вообще не БЭМ :) Если бы мне нужно было добавить светлую тему, я бы сделал:

```
.theme--light
.theme--dark
```

Не очень понятно, зачем именно писать `color`, это ведь и так тема, там цвета :) Ну, кроме привычки.

Но если было бы важно разделить текст и фон:

```
.theme--text-light
.theme--back-dark
```

**Reply by pepelsbey** [reply]
@monochromer можешь вкинуть реальные примеры в двух разных нотациях? Мы бы их сравнили.

**Reply by monochromer** [reply]
> это вообще не БЭМ :) 

По ссылке выше вроде официальная документация )

> зачем именно писать color

Есть тема для цвета, есть тема для шрифта, ещё могла быть тема для отступов... - `theme_color_light`, `theme_font_code`

Когда мы применяем один булевый модификатор, то точно не знаем, является ли он взаимоисключающим. То есть я точно знаю, что должен быть либо `theme_color_light`, либо `theme_color_dark`. А если писать так `theme--light` и `theme--font-code`, то должен ли я удалить `theme--font-code` чтобы поставить `theme--light`?

Ещё примеры: 
```
avatar_size_big
avatar_size_small
avatar_size_medium

avatar_shape_round
avatar_shape_rect
```
Как они должны быть записаны в `--`-нотации? Так, наверно:
```
avatar--size--big
avatar--size--small
avatar--size--medium

avatar--shape--round
avatar--shape--rect
```

**Reply by furtivite** [reply]
> зачем именно писать `color`

@pepelsbey это модификатор ключ-значение, в CSS-то конечно понятно, что меняется свойство цвета, а не заливки, но если убрать `color` будет не ясно в HTML там, где это применяется.

> avatar--size--big
avatar--size--small
avatar--size--medium

avatar--shape--round
avatar--shape--rect

Кажется, это какой-то перепридуманный Two Dashes. 

Должно быть так:

```css
avatar__size--big
avatar__size--small
avatar__size--medium

avatar__shape--round
avatar__shape--rect
```

**Reply by monochromer** [reply]
нет, здесь получается `shape` и `rect` элементы, а должны быть именами модификаторов

**Reply by pepelsbey** [reply]
> По ссылке выше вроде официальная документация )

По ссылке размышления в стиле «мы так не пишем, но если выражать ключ-значение, то будет вот такой нонсенс». Так никто не пишет, я ни раз не видел такой записи. То есть нет, если нужен модификатор из двух частей (с префиксом, ключом или как там), то он записывается вот так:

```
.avatar--size-big
.avatar--size-small
.avatar--size-medium
```

Главная идея международной нотации в том, что ты явно видишь, где у тебя блок, где элемент, где модификатор.

То есть:

```
.block__element
.block--modifier
```

И если модификатор двухчастный, то вот так:

```
.block--key-value
```

В нотации Яндекса у нас получается змейка

```
.block_key_value
```

Более того, начинаются приключения «мне ключ-значение не нужно, но придётся записать».

Я как раз и хотел попросить показать пример, где ключ-значение на самом деле помогают. И если блок настолько сложный, что у него батарея префиксов, среди которых сложно выделить группы, то может быть блок переусложнён?

К сожалению, у нас скорее спор привычек. Скажу честно, в такой записи:

```
.theme_color_light
```

Я вижу три равнозначных слова, которые не говорят мне о каких-то связях между собой. А вот здесь:

```
.theme--light
```

Я вижу светлую тему. Поэтому мне кажется, что всё можно решить только сбором мнений на основе коллекции примеров.

**Reply by monochromer** [reply]
Важно относится к БЭМ, как к не просто методологии именования CSS-классов, а как к проекции состояния и атрибутов компонентов на визуальное оформление. То есть CSS-класс в БЭМ говорит не о том, как выглядит компонент, а в каком состоянии он находится и какие свойства у него есть.

Если попробовать записать БЭМ-компоненты в XML-стиле, то мы можем получить что-то такое:
```xml
<Theme color="dark" font="code">
</Theme>
```
Здесь должна быть сразу видна необходимость в нотации "ключ-значение". Также видно, что у атрибута может быть только одно значение, ведь вы же не будете писать так: `<Theme color="dark" color="light">`, так как не понятно, что применится. Но вот с "булевской" нотацией все уже не так очевидно: `<Theme dark light>`, `<Theme color-light color-dark>` или может так:
```xml
<Theme dark>
  <Theme light>
  </Theme>
</Theme>
```
Поэтому в Яндексе и отказались от нее.

Ещё пример (не имеет отношения к стилизации, но смысл должен быть понятен): как бы задавали атрибуты `loading` и его значения в разных нотациях для тега `img`:
```html
<img loading="lazy">
<img loading="eager">
```
или
```html
<img loading-lazy>
<img loading-eager>
```
Что проще понять?

**Reply by pepelsbey** [reply]
> Важно относится к БЭМ, как к не просто методологии именования CSS-классов

Мне кажется, что тут не «важно», а «можно» ;) Если мы говорим про модификаторы простых блоков и элементов, то в моём опыте всегда хватает простых модификаторов, ключ-значение не нужно. А если не нужно, то зачем усложнять код?

Но тут снова нужно вернуться к тому, что я сказал выше:

> К сожалению, у нас скорее спор привычек

**Reply by furtivite** [reply]
@pepelsbey можешь, на всякий случай (кажется я запутался, потому что у редко встречаю неяндексовую нотацию), как ты предлагаешь записывать:

- блок (у Яндекса так: `.block`)
- блок с модификатором (у Яндекса так: `.block_modifier`)
- элемент (у Яндекса так: `.block__element`)
- элемент с модификатором (у Яндекса так: `.block__element_modifier`)

**Reply by pepelsbey** [reply]
У меня созрел аргумент :) Представьте вот такой пример:

```html
<tag lazy></tag>
```

Как и в XML-примерах выше, не очень понятно, что здесь происходит. Мы не знаем, что за тег, что это за лейзи. Но как только мы видим такое:

```html
<input required>
<main hidden>
```

Вопросов уже нет: мы понимаем, что это значит по тегу. То же самое с блоками: если блок достаточно атомарен и понятен как input, img или main, то бинарные модификаторы прекрасно работают. Он либо есть, либо его нет — а то, что он делает, перечислено в ряду немногих модификаторов.

**Reply by furtivite** [reply]
> если блок достаточно атомарен и понятен как input, img или main, то бинарные модификаторы прекрасно работают. Он либо есть, либо его нет — а то, что он делает, перечислено в ряду немногих модификаторов.

Я соглашусь, что надо сделать так, что бы блок (или элемент) был максимально понятен, а модификатор был бы скорее исключением, чем правилом. Вопрос, в том, что в дальнейшем нам может понадобиться модификатор формата ключ-значение, а если мы сейчас скажем, что такого быть не может и мы так не делаем, то это повлияет на то, что будет в проекте дальше.

**Comment by igsekor** [comment]
Кажется, что европейская нотация легче воспринимается. Согласен с @pepelsbey. Кажется, что перегружать название классов и переменных не стоит. Да и зачем? Всё же и так понятно.

А зачем разделять понятия тема для шрифта и тема для цветов? Кажется, что слово theme в `theme_color_light`, `theme_font_code` в названии двух классов лишнее. Разве нет? Имена классов `theme--light` и `font--code` не понятнее ли (если `font` употребляется для общих настроек размеров или цвета, например)? У меня готового решения нет, но кажется, что много слов в некоторых частях нашего CSS является избыточным.

В то же время, пользоваться возможностями нотации Яндекса в тех случаях, когда это избыточно, но пользоваться приятно — это вкусовщина? Так же, как и желание использовать международную нотацию.

Плюсы международной нотации в том, что её используют за пределами Яндекса. Разве нет? Может нам важно делать так, чтобы ассоциироваться с сообществом даже в коде, а не с Яндексом, нотацию которого используют в основном в русскоязычном сообществе, да и то далеко не везде?

**Reply by igsekor** [reply]
Вот список всех классов не данный момент в нотации Яндекса из [PR](https://github.com/doka-guide/platform/pull/275):

```css
.base-list

.base
.base__body

.footer-nav-list
.footer-nav-list__item
.footer-nav-list__link

.footer
.footer__toggle
.footer__list

.link

.theme-toggle
.theme-toggle__input
.theme-toggle__desc
.theme-toggle__control
.theme-toggle__thumb
.theme-toggle__icon
.theme-toggle__icon_light
.theme-toggle__icon_dark

.theme
.theme_font_text
.theme_font_code
.theme_color_light
.theme_color_dark

.visually-hidden
```

Может каждый посмотрит на этот список и предложит свой вариант именования или согласится с тем, который уже есть?

**Comment by pepelsbey** [comment, marked as answer]
Предлагаю использовать следующую нотацию. См. дискуссию выше.

```
.block
.block--modifier
.block__element
.block__element--modifier
```

@furtivite вот так ^

**Comment by monochromer** [comment]
Как насчёт использования data-атрибутов в качестве модификаторов?
```
:root[data-dark-theme] {}

.avatar[data-size="small"] {}
.avatar[data-size="medium"] {}
.avatar[data-shape="round"] {}
.avatar[data-shape="rect"] {}

.dropdown[data-state="open"] {}
```

**Reply by pepelsbey** [reply]
Они повышают вес селектора и добавляют шума — читать сложнее. Мне не очень нравится этот подход.

**Reply by monochromer** [reply]
Зато они как раз выравнивают специфичность, когда используются рядом с псевдоклассами типа `:checked`, `:hover`, `:focus`

**Reply by furtivite** [reply]
> Зато они как раз выравнивают специфичность, когда используются рядом с псевдоклассами типа `:checked`, `:hover`, `:focus`

Выровнять при необходимости можно с указанием родительского блока `.block .block__element`, а где это не требуется, там и выравнивать нечего.

**Reply by monochromer** [reply]
Это `.block .block__element` противоречит методологии

**Reply by furtivite** [reply]
> Это `.block .block__element` противоречит методологии

Методология [допускает](https://ru.bem.info/methodology/css/#вложенные-селекторы), хотя и не рекомендует вложенные селекторы. 

Это же не комбинированный селектор, и его можно легко переиспользовать внутри блока. Не будет проблем.

**Reply by monochromer** [reply]
Там написано, что можно с помощью модификаторов на "родителе" менять "детей"
